datastore/filesystem/filesystem.py

```diff
@@ -1,8 +1,87 @@
+import errno
+import io
 import os
-from datastore.abc import BinaryDatastore
-
-
-class FileSystemDatastore(BinaryDatastore):
+import typing
+
+import trio
+
+import datastore
+import datastore.abc
+
+from .util import statx
+
+
+# Make default buffer larger to try to compensate for the thread switching overhead
+DEFAULT_BUFFER_SIZE = io.DEFAULT_BUFFER_SIZE * 10
+
+
+stat_result_t = typing.Union[os.stat_result, statx.stat_result]
+
+
+class FileReader(datastore.abc.ReceiveStream):
+	__slots__ = ("_file")
+	
+	_file: 'trio._file_io.AsyncIOWrapper'
+	
+	
+	def __init__(self, file: 'trio._file_io.AsyncIOWrapper', stat: stat_result_t):
+		super().__init__()
+		
+		self._file = file
+		
+		self.size  = stat.st_size
+		self.atime = stat.st_atime
+		self.mtime = stat.st_mtime
+		if stat.st_atime_ns:
+			self.atime = stat.st_atime_ns / 1_000_000_000
+		if stat.st_mtime_ns:
+			self.mtime = stat.st_mtime_ns / 1_000_000_000
+		
+		# Finding btime from stat is tricky and platform dependant
+		st_birthtime_ns: typing.Optional[int] = getattr(stat, "st_birthtime_ns", None)
+		if st_birthtime_ns:
+			# Linux with statx patch exposes this
+			self.btime = st_birthtime_ns / 1_000_000_000
+		elif hasattr(stat, "st_birthtime") and stat.st_birthtime:
+			# FreeBSD/macOS has this field
+			self.btime = stat.st_birthtime
+		elif os.name == "nt":  # Windows stores btime as ctime
+			self.btime = stat.st_ctime
+			if stat.st_ctime_ns:
+				self.btime = stat.st_ctime_ns / 1_000_000_000
+	
+	
+	async def receive_some(self, max_bytes: typing.Optional[int] = None):
+		if max_bytes:
+			buf = await self._file.read(max_bytes)
+		else:
+			buf = await self._file.read(DEFAULT_BUFFER_SIZE)
+		
+		if len(buf) == 0:
```

Comment:
Is it necessary to call .aclose() here? Won't the user explicitly call it?
And if we are already calling .aclose(), when the user explicitly calls .aclose(), wont there be some kind of error, saying that it is already closed?

Reply:
Calling `.aclose()` here ensures that the file automatically be closed in an

```py
async for chunk in await datastore.get("whatever"):
    …
```

kind of scenario. Otherwise the calling code would have to look like this:

```py
async with await datastore.get("whatever") as file:
    async for chunk in file:
        …
```

… or the file descriptor would leak.

Calling `aclose` several times [is not an error](https://trio.readthedocs.io/en/stable/reference-io.html#trio.abc.AsyncResource.aclose):

> If the resource is already closed, then this method should silently succeed.